src/daemon/distribution.c

```diff
@@ -0,0 +1,262 @@
+    free(buckets);
+    return NULL;
+  }
+  new_distribution->buckets = buckets;
+
+  for (size_t i = 0; i < num_bounds + 1; i++) {
+    if (i == 0) {
+      new_distribution->buckets[i] =
+          initialize_bucket(0, custom_max_boundaries[0]);
+    } else if (i < num_bounds) {
+      new_distribution->buckets[i] =
+          initialize_bucket(new_distribution->buckets[i - 1].max_boundary,
+                            custom_max_boundaries[i]);
+    } else {
+      new_distribution->buckets[i] = initialize_bucket(
+          new_distribution->buckets[i - 1].max_boundary, INFINITY);
+    }
+  }
+
+  new_distribution->num_buckets =
+      num_bounds + 1; // plus one for infinity bucket
+  new_distribution->total_scalar_count = 0;
+  new_distribution->raw_data_sum = 0;
+  return new_distribution;
+}
+
+static size_t binary_search(distribution_t *dist, size_t left, size_t right,
+                            double gauge) {
+  if (left > right) {
+    return -1;
+  }
+
+  size_t mid = left + (right - left) / 2;
+  if (gauge >= dist->buckets[mid].min_boundary &&
+      gauge < dist->buckets[mid].max_boundary) {
+    return mid;
+  }
+
+  if (gauge < dist->buckets[mid].min_boundary) {
+    return binary_search(dist, left, mid, gauge);
+  }
+
+  return binary_search(dist, mid, right, gauge);
+}
+
+void distribution_update(distribution_t *dist, double gauge) {
+
+  size_t left = 0;
+  size_t right = dist->num_buckets - 1;
+  size_t index = binary_search(dist, left, right, gauge);
+  /*Linear update
+  for(size_t i = 0; i < dist->num_buckets; i++) {
+      if(gauge >= dist->buckets[i].min_boundary && gauge <=
+  dist->buckets[i].max_boundary) { dist->buckets[i].bucket_counter++;
+      }
+  }*/
+  dist->buckets[index].bucket_counter++;
+  dist->total_scalar_count++;
+  dist->raw_data_sum += gauge;
+}
+
+double distribution_average(distribution_t *dist) {
+  return dist->raw_data_sum / dist->total_scalar_count;
+}
+
+double distribution_percentile(distribution_t *dist, double percent) {
+  if ((percent < 0) || (percent > 100)) {
+    errno = EINVAL;
+    return NAN;
+  }
+
+  int sum = 0;
+  double bound = 0;
+  double target_amount = (percent * 100) / dist->num_buckets;
+  for (size_t i = 0; i < dist->num_buckets; i++) {
+    sum += dist->buckets[i].bucket_counter;
+    if ((double)sum >= target_amount) {
+      bound = dist->buckets[i].max_boundary;
+      break;
+    }
+  }
+  return bound;
+}
+
+distribution_t *distribution_clone(distribution_t *dist) {
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  bucket_t *buckets = calloc(dist->num_buckets, sizeof(bucket_t));
+
+  if ((new_distribution == NULL) || (buckets == NULL)) {
+    free(new_distribution);
+    free(buckets);
+    return NULL;
+  }
+  new_distribution->buckets = buckets;
+  new_distribution->num_buckets = dist->num_buckets;
+
+  for (size_t i = 0; i < new_distribution->num_buckets; i++) {
+    new_distribution->buckets[i] = initialize_bucket(
```

Comment:
I think this is missing the counter stored in each bucket.

You can use the [memcpy(3)](https://man7.org/linux/man-pages/man3/memcpy.3.html) standard function to copy a block of memory (i.e. all the buckets) from one location to another one:

```c
memcpy(new_distribution->buckets, dist->buckets, sizeof(bucket_t) * dist->num_buckets);
```

Reply:
Thank you, I am using memcpy now.